File: benchmarks/programs/cpp/hashmap.cpp
// Benchmark: unordered_map insertion and lookup throughput.
// n=200_000 — identical across Skarn, Python, C++.
// No reserve() so all three languages pay the same rehash/growth cost inside the timer.
#include <cstdio>
#include <chrono>
#include <unordered_map>

int main() {
    const int n = 200'000;

    auto t0 = std::chrono::high_resolution_clock::now();

    std::unordered_map<int, int> m;

    for (int i = 0; i < n; ++i) m[i] = i % 1000;

    long long checksum = 0;
    for (int i = 0; i < n; ++i) {
        auto it = m.find(i);
        if (it != m.end()) checksum += it->second;
    }

    auto wall_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::high_resolution_clock::now() - t0).count();

    std::printf("benchmark=hashmap  n=%d  checksum=%lld  wall_ns=%lld  ns/op=%lld\n",
                n, checksum, wall_ns, wall_ns / (2LL * n));
    std::printf("gc_collections=0  gc_bytes=0  gc_ns=0\n");
    return 0;
}

File: benchmarks/programs/python/hashmap.py
#!/usr/bin/env python3
# Benchmark: dict insertion and lookup throughput.
# n=200_000 — identical across Skarn, Python, C++.
import time, gc

_gc_ns = 0; _gc_start = 0
def _gc_cb(phase, info):
    global _gc_ns, _gc_start
    if phase == 'start': _gc_start = time.perf_counter_ns()
    else: _gc_ns += time.perf_counter_ns() - _gc_start
gc.callbacks.append(_gc_cb)

def main():
    gc_before = sum(s['collections'] for s in gc.get_stats())
    n = 200_000
    d = {}
    t0 = time.perf_counter_ns()
    for i in range(n):
        d[i] = i % 1000
    checksum = 0
    for i in range(n):
        checksum += d.get(i, 0)
    wall_ns = time.perf_counter_ns() - t0
    gc_collections = sum(s['collections'] for s in gc.get_stats()) - gc_before
    print(f"benchmark=hashmap  n={n}  checksum={checksum}  wall_ns={wall_ns}  ns/op={wall_ns // (n * 2)}")
    print(f"gc_collections={gc_collections}  gc_bytes=0  gc_ns={_gc_ns}")

if __name__ == "__main__":
    main()
